types/currency_test.go

```diff
@@ -0,0 +1,205 @@
+package types
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"testing"
+)
+
+// TestCurrencyToBig tests the Big method for the currency type
+func TestCurrencyToBig(t *testing.T) {
+	c := NewCurrency64(125)
+	cb := c.Big()
+	b := big.NewInt(125)
+
+	if b.Cmp(cb) != 0 {
+		t.Error("currency to big has failed")
+	}
+}
+
+// TestCurrencyDiv checks that the div function has been correctly implemented.
+func TestCurrencyDiv(t *testing.T) {
+	c9 := NewCurrency64(9)
+	c10 := NewCurrency64(10)
+	c90 := NewCurrency64(90)
+	c97 := NewCurrency64(97)
+
+	c90D10 := c90.Div(c10)
+	if c90D10.Cmp(c9) != 0 {
+		t.Error("Dividing 90 by 10 should produce 9")
+	}
+	c97D10 := c97.Div(c10)
+	if c97D10.Cmp(c9) != 0 {
+		t.Error("Dividing 97 by 10 should produce 9")
+	}
+}
+
+// TestCurrencySqrt checks that the sqrt function of the currency type has been
+// correctly implemented.
+func TestCurrencySqrt(t *testing.T) {
+	c8 := NewCurrency64(8)
+	c64 := NewCurrency64(64)
+	c80 := NewCurrency64(80)
+	sqrt64 := c64.Sqrt()
+	sqrt80 := c80.Sqrt()
+
+	if c8.Cmp(sqrt64) != 0 {
+		t.Error("square root of 64 should be 8")
+	}
+	if c8.Cmp(sqrt80) != 0 {
+		t.Error("square root of 80 should be 8")
+	}
+}
+
+// TestCurrencyMarshalJSON probes the MarshalJSON and UnmarshalJSON functions
+// of the currency type.
+func TestCurrencyMarshalJSON(t *testing.T) {
+	b30 := big.NewInt(30)
+	c30 := NewCurrency64(30)
+
+	bMar30, err := b30.MarshalJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+	cMar30, err := c30.MarshalJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if bytes.Compare(bMar30, cMar30) != 0 {
+		t.Error("Currency does not match the marshalling of its math/big equivalent")
+	}
+
+	var cUmar30 Currency
+	err = cUmar30.UnmarshalJSON(cMar30)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if c30.Cmp(cUmar30) != 0 {
+		t.Error("Incorrect unmarshalling of currency type.")
+	}
+
+	cMar30[0] = 0
+	err = cUmar30.UnmarshalJSON(cMar30)
+	if err == nil {
+		t.Error("JSON decoded nonsense input")
+	}
+}
+
+// TestCurrencyMarshalSia probes the MarshalSia and UnmarshalSia functions of
+// the currency type.
+func TestCurrencyMarshalSia(t *testing.T) {
+	c := NewCurrency64(1656)
+	cMar := c.MarshalSia()
+	var cUmar Currency
+	cUmar.UnmarshalSia(cMar)
+	if c.Cmp(cUmar) != 0 {
+		t.Error("marshal and unmarshal mismatch for currency type")
+	}
+}
+
+// TestCurrencyString probes the String function of the currency type.
+func TestCurrencyString(t *testing.T) {
+	b := big.NewInt(7135)
+	c := NewCurrency64(7135)
+	if b.String() != c.String() {
+		t.Error("string function not behaving as expected")
+	}
```

Comment:
I actually caught an error in one of the pass-through functions, so it's not completely silly.